### PR TITLE
DOC-1160 - RS API: Document v1 /actions/ API request.

### PR DESCRIPTION
**Pull request by cmilesb**
Staging Links: 

Actions endpoint: https://docs.redis.com/staging/jira-doc-1160/rs/references/rest-api/requests/actions/
State Machine object: https://docs.redis.com/staging/jira-doc-1160/rs/references/rest-api/objects/state-machine/

Draft of the /actions and /actions/{uid} GET requests.

### CODE REVIEW

File: content/rs/references/rest-api/objects/state-machine.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+---
+Title: State Machine object
+linkTitle: state-machine
+description: An object that represents state machines.
+weight: $weight
+alwaysopen: false
+categories: ["RS"]
+---
+
+A `state-machine` object tracks the creation of a database on a Redis Enterprise cluster. When you create a database on a Redis Enterprise cluster, Redis Enterprise creates a state machine that tracks the status of the database creation process.
```
Comment (lanceleonard) [inline content/rs/references/rest-api/objects/state-machine.md:10]: Consider linking "state machine" to the API request lifecycle doc.

File: content/rs/references/rest-api/requests/actions.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+| [GET](#get-all-actions) | `/v1/actions` | Get all actions |
+| [GET](#get-action) | `/v1/actions/{uid}` | Get a single action |
+
+## Get all actions {#get-all-actions}
+
+```
+GET /v1/actions
+```
+
+Get the status of all currently executing, queued, or completed actions on all entities (clusters, nodes, and databases). This API is used to track every long-lived API request which returns either a `task_id` or an `action_uid`.
```
Comment (lanceleonard) [inline content/rs/references/rest-api/requests/actions.md:28]:
```suggestion
Get the status of all actions (executing, queued, or completed) on all entities (clusters, nodes, and databases). This API tracks long-lived API requests return either a `task_id` or an `action_uid`.
```

File: content/rs/references/rest-api/requests/actions.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    }
+  ]
+}
+```
+
+### Status codes {#get-all-status-codes}
+
+| Code | Description |
+|------|-------------|
+| [200 OK](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.1) | No error, response provides info about an ongoing action. |
```
Comment (lanceleonard) [inline content/rs/references/rest-api/requests/actions.md:87]:
```suggestion
| [200 OK](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.1) | No error, response provides info about an ongoing action |
```

File: content/rs/references/rest-api/requests/actions.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+| [404 Not Found](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.5) | Action does not exist (i.e. not currently running and no available status of last run). |
```
Comment (lanceleonard) [inline content/rs/references/rest-api/requests/actions.md:88]:
```suggestion
| [404 Not Found](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.5) | Action does not exist (i.e. not currently running and no available status of last run).|
```

File: content/rs/references/rest-api/requests/actions.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+```
+GET /actions/{uid}
+```
+
+#### URL parameters
+
+| Field | Type | Description |
+|-------|------|-------------|
+| uid | string | The action_uid to check. |
```
Comment (lanceleonard) [inline content/rs/references/rest-api/requests/actions.md:116]:
```suggestion
| uid | string | The action_uid to check |
```

File: content/rs/references/rest-api/requests/actions.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+  "status": "completed",
+  "task_id": "159ca2f8-7bf3-4cda-97e8-4eb560665c28"
+}
+```
+
+### Status codes {#get-status-codes}
+
+| Code | Description |
+|------|-------------|
+| [200 OK](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.1) | No error, response provides info about an ongoing action. |
```
Comment (lanceleonard) [inline content/rs/references/rest-api/requests/actions.md:141]:
```suggestion
| [200 OK](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.1) | No error, response provides info about an ongoing action |
```

File: content/rs/references/rest-api/requests/actions.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+| [404 Not Found](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.5) | Action does not exist (i.e. not currently running and no available status of last run). |
```
Comment (lanceleonard) [inline content/rs/references/rest-api/requests/actions.md:142]:
```suggestion
| [404 Not Found](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.5) | Action does not exist (i.e. not currently running and no available status of last run) |
```

**Review by lanceleonard — APPROVED** [review]
LGTM, with some recommendations to consider.

File: content/rs/references/rest-api/objects/state-machine.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+A state machine contains the following attributes:
+
+| Name        | Type/Value | Description |
+|-------------|------------|-------------|
+| action_uid  | string     | A globally unique identifier of the action |
+| object_name | string     | Name of the object being manipulated by the state-machine |
+| status      | pending    | Requested state machine has not started |
+|             | active     | State machine is currently running |
+|             | completed  | Operation complete |
+|             | failed     | Operation or State Machine failed |
```
Comment (nermiller) [inline content/rs/references/rest-api/objects/state-machine.md:21]: consider using `state-machine` here as well. just checking: this and the next two statuses don't have a name?

File: content/rs/references/rest-api/requests/actions.md (hunk 2 above)
Comment (cmilesb) [reply content/rs/references/rest-api/requests/actions.md:28]:
```suggestion
Get the status of all actions (executing, queued, or completed) on all entities (clusters, nodes, and databases). This API tracks long-lived API requests that return either a `task_id` or an `action_uid`.
```
Made a minor change to the second sentence here - let me know what you think.

File: content/rs/references/rest-api/objects/state-machine.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
+---
+Title: State Machine object
```
Comment (rrelledge) [inline content/rs/references/rest-api/objects/state-machine.md:2]:
```suggestion
Title: State machine object
```

File: content/rs/references/rest-api/requests/actions.md [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+| Field | Type | Description |
+|-------|------|-------------|
+| uid | string | The action_uid to check. |
+
+### Response {#get-response}
+
+Returns an [action object]({{<relref "/rs/references/rest-api/objects/action">}}).
+
+Regardless of an action’s source, each action contains the following attributes: `name`, `action_uid`, `status` and `progress`.
```
Comment (nermiller) [inline content/rs/references/rest-api/requests/actions.md:122]: add , before _and_

**Review by nermiller — APPROVED** [review]
Just a couple of nits, Cameron. Nice job writing up these pages! Clear and concise!

File: content/rs/references/rest-api/objects/state-machine.md (hunk 1 above)
Comment (rrelledge) [inline content/rs/references/rest-api/objects/state-machine.md:10]: Can you streamline this paragraph?

File: content/rs/references/rest-api/requests/actions.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+| Field | Type | Description |
+|-------|------|-------------|
+| uid | string | The action_uid to check |
+
+### Response {#get-response}
+
+Returns an [action object]({{<relref "/rs/references/rest-api/objects/action">}}).
+
+Regardless of an action’s source, each action contains the following attributes: `name`, `action_uid`, `status` and `progress`.
```
Comment (cmilesb) [inline content/rs/references/rest-api/requests/actions.md:122]:
```suggestion
Regardless of an action’s source, each action contains the following attributes: `name`, `action_uid`, `status`, and `progress`.
```

File: content/rs/references/rest-api/objects/state-machine.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+linkTitle: state-machine
+description: An object that represents state machines.
```
Comment (rrelledge) [inline content/rs/references/rest-api/objects/state-machine.md:4]:
```suggestion
description: An object that represents a state machine.
```

File: content/rs/references/rest-api/objects/state-machine.md (hunk 8 above)
Comment (cmilesb) [reply content/rs/references/rest-api/objects/state-machine.md:21]: I'm using the table here similarly to the Action object - https://docs.redis.com/staging/jira-doc-1160/rs/references/rest-api/objects/action/ . The "status" can be either pending, active, completed, or failed. I'm using the "Description" column to describe each state.

Comment (rrelledge) [inline content/rs/references/rest-api/objects/state-machine.md:21]: Also, maybe change the other occurrences of "state-machine" to "state machine" to have a consistent spelling.
```suggestion
|             | failed     | Operation or state machine failed |
```

File: content/rs/references/rest-api/requests/actions.md [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+#### Example HTTP request
+
+```
+GET /actions
+```
+
+### Response {#get-all-response}
+
+Returns a JSON array of [action objects]({{<relref "/rs/references/rest-api/objects/action">}}) and an array of [state-machine objects]({{<relref "/rs/references/rest-api/objects/state-machine">}}). Regardless of an action’s source, each action in the response contains the following attributes: `name`, `action_uid`, `status` and `progress`.
```
Comment (rrelledge) [inline content/rs/references/rest-api/requests/actions.md:46]: Maybe add a line break and a missing comma.
```suggestion
Returns a JSON array of [action objects]({{<relref "/rs/references/rest-api/objects/action">}}) and an array of [state-machine objects]({{<relref "/rs/references/rest-api/objects/state-machine">}}).

Regardless of an action’s source, each action in the response contains the following attributes: `name`, `action_uid`, `status`, and `progress`.
```

**Review by rrelledge — CHANGES_REQUESTED** [review]
Looks good. I just left a few suggestions for changes.

File: content/rs/references/rest-api/requests/actions.md (hunk 10 above)
Comment (cmilesb) [reply content/rs/references/rest-api/requests/actions.md]: Fixed in https://github.com/RedisLabs/redislabs-docs/pull/1885/commits/1b1375a7d0f8a5cd3fa04b826f2752832f91e800

File: content/rs/references/rest-api/objects/state-machine.md (hunk 1 above)
Comment (cmilesb) [reply content/rs/references/rest-api/objects/state-machine.md]: Fixed in https://github.com/RedisLabs/redislabs-docs/pull/1885/commits/e116a51a5dba34fbce7b05b65a861081441f7257.

**Review by rrelledge — APPROVED** [review]
Looks good to me!

**Pull request merged**